clients/drcachesim/tests/drcachesim_unit_tests.cpp

```diff
@@ -306,9 +308,201 @@ LLC {
+void
+unit_test_cache_associativity()
+{
+    // Range of associativities to be tested.
+    constexpr int MIN_ASSOC = 1;
+    constexpr int MAX_ASSOC = 16;
+
+    constexpr int LINE_SIZE = 32;
+    constexpr int BLOCKS_PER_WAY = 16;
+
+    // Only power-of-two associativities are currently supported.
+    for (int assoc = MIN_ASSOC; assoc <= MAX_ASSOC; assoc *= 2) {
+        int total_size = LINE_SIZE * BLOCKS_PER_WAY * assoc;
+        // Test access patterns that stress increasing associativity.
+        for (uint32_t test_assoc = 1; test_assoc <= 2 * assoc; ++test_assoc) {
+            cache_lru_t cache;
+            caching_device_stats_t stats(/*miss_file=*/"", LINE_SIZE);
+            assert(cache.init(assoc, LINE_SIZE, total_size, /*parent=*/nullptr, &stats));
+            // Test start address is arbitrary.
+            addr_t start_address = test_assoc * total_size;
+
+            constexpr int NUM_LOOPS = 3; // Anything >=2 should work.
+            auto read_count =
+                generate_2D_accesses(cache, start_address, LINE_SIZE, BLOCKS_PER_WAY,
+                                     total_size, test_assoc, NUM_LOOPS);
+            CStats c_stats = get_cache_stats(stats);
+            assert(read_count == NUM_LOOPS * BLOCKS_PER_WAY * test_assoc);
+
+            int expected_misses =
+                (test_assoc <= assoc) ? BLOCKS_PER_WAY * test_assoc : read_count;
+            int expected_hits = read_count - expected_misses;
+            if (expected_hits != c_stats.hits || expected_misses != c_stats.misses) {
+                std::cerr << "ERROR in unit_test_cache_associativity():"
+                          << " test_assoc=" << test_assoc << " read_count=" << read_count
+                          << " expected_hits=" << expected_hits
+                          << " actual_hits=" << c_stats.hits
+                          << " expected_misses=" << expected_misses
+                          << " actual_misses=" << c_stats.misses << "\n";
+            }
+            assert(expected_hits == c_stats.hits);
+            assert(expected_misses == c_stats.misses);
+        }
+    }
+}
+
+// Tests an LRU cache to verify it behaves as its size requires.
+void
+unit_test_cache_size()
+{
+    // Range of cache sizes to test.  16KB - 4MB hits the most interesting ones.
+    constexpr int MIN_TOTAL_SIZE = 16 * 1024;
+    constexpr int MAX_TOTAL_SIZE = 4 * 1024 * 1024;
+    constexpr int LINE_SIZE = 64;
+    constexpr int ASSOCIATIVITY = 2;
+
+    for (int cache_size = MIN_TOTAL_SIZE; cache_size <= MAX_TOTAL_SIZE; cache_size *= 2) {
+        // Access a buffer of increasing size, make sure hits + misses are expected.
+        for (int buffer_size = cache_size / 2; buffer_size < cache_size * 2;
+             buffer_size *= 2) {
+            cache_lru_t cache;
+            caching_device_stats_t stats(/*miss_file=*/"", LINE_SIZE);
+            cache.init(ASSOCIATIVITY, LINE_SIZE, cache_size, /*parent=*/nullptr, &stats);
```

Comment:
Seems worth checking the return value.

Reply:
Oops, thought I had fixed all those.  Done.